File: data/src/main/java/com/arturogutierrez/openticator/domain/password/PasswordSerializerImpl.java
package com.arturogutierrez.openticator.domain.password;

import android.util.Base64;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import javax.inject.Inject;

public class PasswordSerializerImpl implements PasswordSerializer {

  @Inject
  public PasswordSerializerImpl() {

  }

  public String encodePassword(String plainPassword) {
    if (plainPassword == null) {
      return null;
    }

    MessageDigest messageDigest = createMessageDigest();
    if (messageDigest == null) {
      return null;
    }

    messageDigest.update(plainPassword.getBytes());
    byte[] digest = messageDigest.digest();
    return Base64.encodeToString(digest, Base64.NO_WRAP).trim();
  }

  private MessageDigest createMessageDigest() {
    try {
      return MessageDigest.getInstance("SHA-512");
    } catch (NoSuchAlgorithmException e) {
      // Should not happen, all Android devices should support SHA-256
      return null;
    }
  }
}

File: data/src/main/kotlin/com/arturogutierrez/openticator/domain/password/PasswordSerializerImpl.kt
package com.arturogutierrez.openticator.domain.password

import android.util.Base64
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import javax.inject.Inject

class PasswordSerializerImpl @Inject constructor() : PasswordSerializer {

  override fun encodePassword(plainPassword: String): String? {
    val messageDigest = createMessageDigest() ?: return null

    messageDigest.update(plainPassword.toByteArray())
    val digest = messageDigest.digest()
    return Base64.encodeToString(digest, Base64.NO_WRAP).trim { it <= ' ' }
  }

  private fun createMessageDigest(): MessageDigest? {
    return try {
      MessageDigest.getInstance("SHA-512")
    } catch (e: NoSuchAlgorithmException) {
      // Should not happen, all Android devices should support SHA-256
      null
    }
  }
}
